client/pluginmanager/csimanager/instance.go

```diff
@@ -86,136 +119,30 @@ func (i *instanceManager) runLoop() {
 		case <-timer.C:
 			ctx, cancelFn := i.requestCtxWithTimeout(managerFingerprintInterval)
 
-			if basicInfo == nil {
-				info, err := i.buildBasicFingerprint(ctx)
-				if err != nil {
-					// If we receive a fingerprinting error, update the stats with as much
-					// info as possible and wait for the next fingerprint interval.
-					info.HealthDescription = fmt.Sprintf("failed initial fingerprint with err: %v", err)
-					cancelFn()
-					i.updater(i.info.Name, basicInfo)
-					timer.Reset(managerFingerprintInterval)
-					continue
-				}
-
-				// If fingerprinting succeeded, we don't need to repopulate the basic
-				// info and we can stop here.
-				basicInfo = info
-			}
-
-			info := basicInfo.Copy()
-			var fp *structs.CSIInfo
-			var err error
-
-			if i.fingerprintNode {
-				fp, err = i.buildNodeFingerprint(ctx, info)
-			} else if i.fingerprintController {
-				fp, err = i.buildControllerFingerprint(ctx, info)
-			}
+			info := i.fp.fingerprint(ctx)
+			cancelFn()
+			i.updater(i.info.Name, info)
 
-			if err != nil {
-				info.Healthy = false
-				info.HealthDescription = fmt.Sprintf("failed fingerprinting with error: %v", err)
+			// TODO: refactor this lock into a faster, goroutine-local check
+			i.volumeManagerMu.RLock()
+			// When we've had a successful fingerprint, and the volume manager is not yet setup,
+			// and one is required (we're running a node plugin), then set one up now.
+			if i.fp.hadFirstSuccessfulFingerprint && !i.volumeManagerSetup && i.fp.fingerprintNode {
```

Comment:
The code for this locking/channel usage looks fine, but it took me a little bit of time with the playground to be sure of myself on that. Rather than having a lock + 2 "completion channels" to reason through, could we reduce the potential for future bugs here by changing the `fingerprint` to block until it's complete (and then immediately have it set up the volume manager for the first time)?

Reply:
the goal is mostly to keep the fingerprinter to be dumb about most things, but I definitely want to simplify this a lot. Probably by having some nicer callback mechanism because I think we'll need a few more things that get triggered on first-success.